core/src/layers/madsim.rs

```
@@ -0,0 +1,318 @@
+use std::fmt::Debug;
+use std::io::SeekFrom;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+#[derive(Debug, Copy, Clone)]
+pub struct MadsimLayer {
+    service_node_id: NodeId,
+}
+
+impl MadsimLayer {
+    pub fn new(service_node_id: NodeId) -> Self {
+        Self { service_node_id }
+    }
+}
+
+impl<A: Accessor> Layer<A> for MadsimLayer {
+    type LayeredAccessor = MadsimAccessor;
+
+    fn layer(&self, _inner: A) -> Self::LayeredAccessor {
+        let handle = simulator::<ServiceSim>().get_node(self.service_node_id);
+        MadsimAccessor { handle }
+    }
+}
+
+#[derive(Debug)]
+pub struct MadsimAccessor {
+    handle: ServiceNodeHandle,
+}
+
+#[async_trait]
+impl LayeredAccessor for MadsimAccessor {
+    type Inner = ();
+    type Reader = MadsimReader;
+    type BlockingReader = MadsimReader;
+    type Writer = MadsimWriter;
+    type BlockingWriter = MadsimWriter;
+    type Pager = MadsimPager;
+    type BlockingPager = MadsimPager;
+
+    fn inner(&self) -> &Self::Inner {
+        &()
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> crate::Result<(RpRead, Self::Reader)> {
+        let handle = self.handle.inner.lock().unwrap();
+        let length = handle
+            .get(path)
+            .map(|(_, data)| data.len())
+            .unwrap_or_default();
+        Ok((
+            RpRead::new(length as u64),
+            MadsimReader {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> crate::Result<(RpWrite, Self::Writer)> {
+        Ok((
+            RpWrite::new(),
+            MadsimWriter {
+                path: path.to_string(),
+                args,
+                handle: self.handle.clone(),
+            },
+        ))
+    }
+
+    async fn list(&self, path: &str, args: OpList) -> crate::Result<(RpList, Self::Pager)> {
+        todo!()
```

Please forward to inner instead of `todo!()`

---

Also, we can return an unsupported errors here. And implement this feature in the future.